src/thunderbots/software/util/canvas_messenger/canvas_messenger.cpp

```diff
@@ -1,139 +1,257 @@
+    const std::chrono::time_point<std::chrono::system_clock> now =
+        std::chrono::system_clock::now();
+    const int64_t elapsed_ns =
+        std::chrono::duration_cast<std::chrono::nanoseconds>(now - time_last_published)
+            .count();
+    const double elapsed_ms = elapsed_ns / 1.0e6;
+
+    // Do not do anything if the time passed hasn't been
+    // long enough
+    if (elapsed_ms < DESIRED_PERIOD_MS)
+        return;
+
+    // Send a payload per layer of messages
+    for (const auto& layer_pair : this->layers_map)
     {
-        this->publisher = node_handle.advertise<thunderbots_msgs::CanvasLayer>(
-            Util::Constants::VISUALIZER_DRAW_LAYER_TOPIC, BUFFER_SIZE);
+        // First is the layer number
+        const uint8_t layer_number = (uint8_t)layer_pair.first;
+
+        // Second is the vector that contains the sprites
+        const std::vector<Sprite>& sprites = layer_pair.second;
+
+        this->publishPayload(layer_number, sprites);
     }
 
-    void CanvasMessenger::publishAndClearLayers()
+    // Clear shapes in layers of current frame/tick
+    this->clearAllLayers();
+
+    // Update last published time
+    time_last_published = now;
+}
+
+void CanvasMessenger::publishPayload(uint8_t layer, std::vector<Sprite> sprites)
+{
+    std::vector<uint8_t> payload;
+
+    // Add the layer number at the begining of the binary message
+    payload.insert(payload.end(), layer);
+
+    // Convert each sprite into a binary message
+    for (Sprite& sprite : sprites)
     {
-        // Limit rate of the message publishing
-        // Get the time right now
-        const std::chrono::time_point<std::chrono::system_clock> now =
-            std::chrono::system_clock::now();
-        const int64_t elapsed_ns = std::chrono::duration_cast<std::chrono::nanoseconds>(
-                                       now - time_last_published)
-                                       .count();
-        const double elapsed_ms = elapsed_ns / 1.0e6;
-
-        // Do not do anything if the time passed hasn't been
-        // long enough
-        if (elapsed_ms < DESIRED_PERIOD_MS)
-            return;
-
-        // Send a payload per layer of messages
-        for (const std::pair<uint8_t, std::vector<Sprite>>& layer_pair : this->layers_map)
-        {
-            // First is the layer number
-            const uint8_t layer_number = layer_pair.first;
-
-            // Second is the vector that contains the sprites
-            const std::vector<Sprite>& sprites = layer_pair.second;
-
-            // Only send the non-empty layers
-            if (!layer_pair.second.empty())
-            {
-                this->publishPayload(layer_number, sprites);
-            }
-        }
-
-        // Clear shapes in layers of current frame/tick
-        this->clearLayers();
-
-        // Update last published time
-        time_last_published = now;
+        std::vector<uint8_t> sprite_payload = sprite.serialize(PIXELS_PER_METER);
```

Comment:
Unfortunately I don't think this will work, because the position values we're giving are in meters and we can't send floating point values in a `uint16_t`?